### PR TITLE
crimson: Support Partial Object Recovery

### PR DESCRIPTION
**Pull request by Matan-B**
## TODO:
  - [x] Track `modified_ranges` interval_set when executing osd ops
  - [x] `OpsExecuter::make_writeable`: Update `clone_overlap` if the most recent clone has been evicted
  - [x] `handle_pull_response()` to `recalc_subsets()`
  - [x] Unblock #49568, #48756
  - [x] `PGBackend::rollback` Calculate clone_overlaps by following overlaps forward from rollback snapshot
  - [x] Add `calc_*_subsets` unit tests and behavior explanation.
## `do_recovery()` Comparisons:
Classic:

Crimson:

* Blocked by: #49568 (Commits from 49568 will be dropped after merge)
* Fixes: https://tracker.ceph.com/issues/58263

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/replicated_recovery_backend.h [hunk 1]
```diff
@@ -47,16 +47,37 @@ class ReplicatedRecoveryBackend : public RecoveryBackend {
...
+    interval_set<uint64_t>& data_subset,
+    std::map<hobject_t, interval_set<uint64_t>>& clone_subsets);
+  void calc_head_subsets(
+    crimson::osd::ObjectContextRef obc,
+    SnapSet& snapset,
+    const hobject_t& head,
+    const pg_missing_t& missing,
+    const hobject_t &last_backfill,
+    interval_set<uint64_t>& data_subset,
+    std::map<hobject_t, interval_set<uint64_t>>& clone_subsets);
```
Comment (athanatos) [inline src/crimson/osd/replicated_recovery_backend.h:67]: Can we just return the data_subset and clone_subsets from calc_*_subsets?  Also, do these actually use any members of ReplicatedRecoveryBackend?  Can we make them free functions in another file and add unit tests for them specifically?

File: src/crimson/osd/replicated_recovery_backend.h [hunk 2]
```diff
@@ -47,16 +47,37 @@ class ReplicatedRecoveryBackend : public RecoveryBackend {
...
+    eversion_t need,
+    std::pair<pg_shard_t, pg_missing_t> pg_shard);
+  void calc_clone_subsets(
+    SnapSet& snapset, const hobject_t& soid,
+    const pg_missing_t& missing,
+    const hobject_t &last_backfill,
+    interval_set<uint64_t>& data_subset,
+    std::map<hobject_t, interval_set<uint64_t>>& clone_subsets);
+  void calc_head_subsets(
+    crimson::osd::ObjectContextRef obc,
```
Comment (athanatos) [inline src/crimson/osd/replicated_recovery_backend.h:61]: Do we need the whole ObjectContext here?  Seems like we only access size through it.

File: src/crimson/osd/replicated_recovery_backend.h (hunk 1 above)
Comment (athanatos) [inline src/crimson/osd/replicated_recovery_backend.h:67]: These two functions encapsulate a mildly complicated amount of logic.  Moreover, their implementations don't actually rely on any members of ReplicatedRecoveryBackend.  Thus, it's worth arranging things to allow us to unit test them in isolation:
- calc_head_subsets doesn't seem to need a full ObjectContextRef, it only seems to access obc->obs.oi.size
- Move to a common header (crimson/osd/object_metadata_helpers.h/cc)
- Return clone_subsets and data_subset in a struct by value.
- Add at least a single unit test for each.

File: src/crimson/osd/replicated_recovery_backend.h [hunk 3]
```diff
@@ -47,16 +47,37 @@ class ReplicatedRecoveryBackend : public RecoveryBackend {
     Ref<MOSDPGRecoveryDelete> m);
   interruptible_future<> handle_recovery_delete_reply(
     Ref<MOSDPGRecoveryDeleteReply> m);
+  interruptible_future<PushOp> prep_push_to_replica(
+    const hobject_t& soid,
+    eversion_t need,
+    std::pair<pg_shard_t, pg_missing_t> pg_shard);
```
Comment (athanatos) [inline src/crimson/osd/replicated_recovery_backend.h:53]: Each of calc_*_subsets needs a comment explaining its behavior.

**Review by athanatos — CHANGES_REQUESTED** [review]
Close!

File: src/crimson/osd/replicated_recovery_backend.h [hunk 4]
```diff
@@ -47,16 +47,37 @@ class ReplicatedRecoveryBackend : public RecoveryBackend {
...
   void prepare_pull(
-    PullOp& po,
-    PullInfo& pi,
+    crimson::osd::ObjectContextRef obc,
+    PullOp& pull_op,
+    pull_info_t& pull_info,
     const hobject_t& soid,
     eversion_t need);
-  std::vector<pg_shard_t> get_shards_to_push(
+  std::map<pg_shard_t, pg_missing_t> get_shards_to_push(
```
[the remark this answers is not in the document]
Comment (Matan-B) [reply src/crimson/osd/replicated_recovery_backend.h:80]: > Alternately, it could be std::map<pg_shard_t, pg_missing_t> including just the entry for the object in question?

`get_shards_to_push()` is returning a vector (`shards`) of `pg_shard_t` according to the following logic:
```
auto shard_missing = pg.get_shard_missing().find(peer);
if (shard_missing->second.is_missing(soid))
  shards.push_back(shard_missing->first);
```
Each of this vector's iterators is passed to `prep_push` immediately (unconditionally) afterwards:
```
const auto& missing = pg.get_shard_missing().find(pg_shard)->second;
const auto it = missing.get_items().find(soid);
assert(it != missing.get_items().end());
```

Instead of pushing only `->first`, and then finding the `pg_shard` again to get `->second`,
`get_shards_to_push()` is updated to return a map and `interruptor::parallel_for_each` will pass pairs to `prep_push` (commit: `Introduce prep_push_to_replica()`).

In short,  we avoid finding the missing peer and asserting it's existence (again) later on in prep_push_to_replica.

[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/osd/replicated_recovery_backend.h:80]: Right, the problem is that it means copying the pg_missing_t structures, which can be big and will involve a lot of allocations.  That's a lot of unnecessary overhead, even for peering.

[the remark this answers is not in the document]
Comment (Matan-B) [reply src/crimson/osd/replicated_recovery_backend.h:80]: Ack, will be fixed.

**Comment by Matan-B** [comment]
Rebased and updated `TODO`. (Still wip)

**Comment by athanatos** [comment]
@Matan-B Is this waiting on a review from me?

**Comment by Matan-B** [comment]
> @Matan-B Is this waiting on a review from me?

Missing unit tests here, I'll add tests and ask for a review.

**Comment by Matan-B** [comment]
@athanatos, ready for review now, thanks!

**Review by athanatos — APPROVED** [review]

**Comment by neha-ojha** [comment]
@jianwei1216 FYI!

**Comment by athanatos** [comment]
Test run looks good: https://pulpito.ceph.com/sjust-2024-01-04_00:41:39-crimson-rados-wip-sjust-crimson-testing-2024-01-03-distro-default-smithi/
